src/main/scala/com/ing/wbaa/rokku/proxy/handler/namespace/RequestHandlerS3WithNamespaces.scala

```diff
@@ -0,0 +1,38 @@
+package com.ing.wbaa.rokku.proxy.handler.namespace
+
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse }
+import com.ing.wbaa.rokku.proxy.data.{ RequestId, S3Request, User }
+import com.ing.wbaa.rokku.proxy.handler.RequestHandlerS3
+import com.ing.wbaa.rokku.proxy.handler.exception.RokkuNamespaceBucketNotFoundException
+
+import scala.concurrent.Future
+
+trait RequestHandlerS3WithNamespaces extends RequestHandlerS3 with NamespacesHandler {
+
+  override protected[this] def executeRequest(request: HttpRequest, userSTS: User, s3request: S3Request)(implicit id: RequestId): Future[HttpResponse] = {
+
+    if (namespaceSettings.isEnabled) {
+      bucketNamespaceCredentials(request) match {
+        case Some(namespaceCredentials) =>
+          val npaRequest = getNpaRequest(request, namespaceCredentials)
```

Comment:
@kr7ysztof Seems that code from line 18-30 is almost identical with the parent super.executeRequest function. The only difference is that this code gets the namespaceCredentials from the caller and the parent super function gets them from the instance context. Seems like an important function since all request will go through this code so it might worth to keep thing DRY here. A quick solution might be to introduce an implicit argument for the super.executeRequest function of type Option[BasicAWSCredentials]. By adjusting the implementation of the function it can first check if the option holds any creds and if not get them from the instance context. I Expect this to not break any existing tests so it should be a low impact change and will deduplicate the existing code. Just thinking out loud here we can also leave it as it is it should be a big deal :)

Reply:
yes DRY is better - but I split the functions - check it.